promql/functions.go

```diff
@@ -79,22 +75,29 @@ func funcDelta(ev *evaluator, args Expressions) model.Value {
 		}
 		resultValue := lastValue - samples.Values[0].Value + counterCorrection
 
-		targetInterval := args[0].(*MatrixSelector).Range
-		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp)
-		if sampledInterval == 0 {
-			// Only found one sample. Cannot compute a rate from this.
-			continue
+		// Duration between first/last samples and boundary of range.
+		durationToStart := samples.Values[0].Timestamp.Sub(rangeStart).Seconds()
+		durationToEnd := rangeEnd.Sub(samples.Values[len(samples.Values)-1].Timestamp).Seconds()
+
+		sampledInterval := samples.Values[len(samples.Values)-1].Timestamp.Sub(samples.Values[0].Timestamp).Seconds()
+		averageDurationBetweenSamples := sampledInterval / float64(len(samples.Values))
+
+		// If the first/last samples are close to the boundaries of the range,
+		// extrapolate the result. This is as we expect that another sample
+		// will exist given the spacing between samples we've seen thus far,
+		// with an allowance for noise.
+		extrapolationThreshold := averageDurationBetweenSamples * 1.1
+		extrpolateToInterval := sampledInterval
+		if durationToStart < extrapolationThreshold {
+			extrpolateToInterval += durationToStart
+		}
+		if durationToEnd < extrapolationThreshold {
+			extrpolateToInterval += durationToEnd
+		}
```

Comment:
I had that other idea of adding _half_ the median duration between samples in cases where we don't extrapolate all the way to the start or end, effectively always extrapolating "a bit" with the rationale that the expectation value for the true start of a series is half a sampling interval before the first sample, and the expectation value for the true end of a series is half a sampling interval after the last sample.